azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java

```diff
@@ -970,6 +992,102 @@ private void cleanOlderProjects() {
         }
       }
     }
+
+    private HashSet<Pair<Integer, Integer>> getActiveProjectVersions() {
+      final HashSet<Pair<Integer, Integer>> activeProjectVersions =
+          new HashSet<>();
+      for (final FlowRunner runner : FlowRunnerManager.this.runningFlows.values()) {
+        final ExecutableFlow flow = runner.getExecutableFlow();
+        activeProjectVersions.add(new Pair<>(flow
+            .getProjectId(), flow.getVersion()));
+      }
+      return activeProjectVersions;
+    }
+
+    private boolean isActiveProject(final ProjectVersion version) {
+      final Pair<Integer, Integer> versionKey = new Pair<>(version.getProjectId(),
+          version.getVersion());
+      return getActiveProjectVersions().contains(versionKey);
+    }
+
+    private boolean shouldCleanup(final File projectDir) {
+      final long dirSize;
+      try {
+        dirSize = FileIOUtils.sizeInKB(projectDir);
+        final long upperLimitInKB = (long) (FlowRunnerManager.this.projectDirMaxSizeInMB *
+            FlowRunnerManager.this.projectDirStartDeletionThreshold * 1024);
+        return dirSize >= upperLimitInKB;
+      } catch (final IOException e) {
+        logger.error(e);
+        return false;
+      }
+    }
+
+    private boolean shouldKeepCleanup(final File projectDir) {
+      final long dirSize;
+      try {
+        dirSize = FileIOUtils.sizeInKB(projectDir);
+        final long lowerLimitInKB = (long) (FlowRunnerManager.this.projectDirMaxSizeInMB *
+            FlowRunnerManager.this.projectDirStopDeletionThreshold * 1024);
+        return dirSize >= lowerLimitInKB;
+      } catch (final IOException e) {
+        logger.error(e);
+        return false;
+      }
+    }
+
+
+    /**
+     * delete least recently used projects
+     */
+    private void cleanLRUProjects() throws IOException {
+      if (shouldCleanup(FlowRunnerManager.this.projectDirectory)) {
+        final List<ProjectVersion> projectVersionList = new ArrayList<>(
+            FlowRunnerManager.this.installedProjects.values());
+
+        // sort project version by last creation time in ascending order
+        projectVersionList.sort((o1, o2) -> {
+          try {
+            final long creationTime1 = creationTime(o1.getInstalledDir());
+            final long creationTime2 = creationTime(o2.getInstalledDir());
+            if (creationTime1 < creationTime2) {
+              return -1;
+            } else if (creationTime1 > creationTime2) {
+              return 1;
+            } else {
+              return 0;
+            }
+          } catch (final IOException ex) {
+            logger.error("error reading project dir", ex);
+            return 0;
+          }
+        });
+
+        for (final ProjectVersion version : projectVersionList) {
```

Comment:
Will it take a long time to calculate the project cache directory size since it's very big? And it will be called inside `shouldKeepCleanup` method which gets executed for every project version in the for loop. Would it be more efficient to do some batch cleaning?

Reply:
yes, but this the frequency of doing it is once/30 mins. 
What do you mean by batch cleaning?

Reply:
I was thinking if it might be possible to clean up like 10 projects at a time and then do the check in `shouldKeepCleanup`. But I'm not sure how much improvement this will have depending on how long it will take in calculating the directory size. If you think efficiency should not be a problem, then I'm ok with this implementation.

Reply:
Have you measured the performance of the current design?

In case performance is proven to be an issue, have you considered the following alternative design? 

Delete older versions of the project immediately. This will free up disk space more quickly.

Create a file in each project directory and write the size of the project to the file when the project is created. The project files are not supposed to change after creation. Touch this file each time the project is used. This way, we can have a more efficient LRU algorithm based on last access time, not creation time. 

Maintain the total size of the project cache in memory to avoid the overhead of re-calculating it. The size shouldn't change too often. This way we can afford to run the check more frequently.